ogr/services/pagure/pull_request.py

```diff
@@ -191,18 +192,43 @@ def get(project: "ogr_pagure.PagureProject", pr_id: int) -> "PullRequest":
         return PagurePullRequest(raw_pr, project)
 
     @staticmethod
-    def get_files_diff(project: "ogr_pagure.PagureProject", pr_id: int) -> dict:
-        try:
-            return project._call_project_api(
-                "pull-request",
-                str(pr_id),
-                "diffstats",
-                method="GET",
-            )
-        except PagureAPIException as ex:
-            if "No statistics" in ex.pagure_error:
-                return {}
-            raise ex
+    def get_files_diff(
+        project: "ogr_pagure.PagureProject",
+        pr_id: int,
+        retries=3,
+        wait_seconds=3,
+    ) -> dict:
+        """Collect all the changes in the PR and deal with ENOPRSTATS exception.
+        While the PR is changing its status from open to merged
+        exception can be raised by the Pagure service.
+        Wait for a while to see if we can manage to access changes adjusting the
+        retries and wait_seconds parameters.
+
+        """
+        attempt = 0
+        while True:
+            try:
+                return project._call_project_api(
+                    "pull-request",
+                    str(pr_id),
+                    "diffstats",
+                    method="GET",
+                )
+            except PagureAPIException as ex:  # noqa PERF203
+                if "No statistics" in ex.pagure_error:
+                    # this may be a race condition, try once more
+                    logger.error(
+                        f"While retrieving PR diffstats Pagure returned ENOPRSTATS. \n{ex}",
+                    )
+                    if attempt <= retries:
+                        logger.error(
+                            f"Trying again; attempt={attempt} after {wait_seconds}seconds",
+                        )
+                        attempt += 1
+                        sleep(wait_seconds)
+                    else:
+                        raise ex
+                raise ex
```

Comment:
I think you are missing `continue` after `sleep()`, otherwise this will always raise.

Reply:
We need a test case for fails first, succeeds when retried.

Reply:
> I think you are missing `continue` after `sleep()`, otherwise this will always raise.

then it complains that the else statement is not usefull :) I think it should be fine like it is now

Reply:
> We need a test case for fails first, succeeds when retried.

We should have both. `test_pr_diff` when the response is 200 and `text_pr_diff_empty_diff` with 400 error

Reply:
> then it complains that the else statement is not usefull :) I think it should be fine like it is now

No, it's not, you would have to add another else above this. But it would be better to `continue` after sleep and remove the else it's complaining about.

Reply:
> > We need a test case for fails first, succeeds when retried.
> 
> We should have both. `test_pr_diff` when the response is 200 and `text_pr_diff_empty_diff` with 400 error

Neither of them tests the complete code and the case I described and this PR is supposed to solve.

Reply:
This is the kind of test I had in mind: https://github.com/majamassarini/ogr/pull/1
It also proves my point about the missing `continue` :slightly_smiling_face:

Reply:
> > then it complains that the else statement is not usefull :) I think it should be fine like it is now
> 
> No, it's not, you would have to add another else above this. But it would be better to `continue` after sleep and remove the else it's complaining about.

I had forgotten an else statement... Now it should be better even without the continue key.

Reply:
Yes, now the logic is correct, however instead of this:

```python
                        sleep(wait_seconds)
                    else:
                        raise ex
                else:
                    raise ex
```
you could have this:

```python
                        sleep(wait_seconds)
                        continue
                raise ex
```

Don't you think it's better?

Reply:
Now I see your point 😄👍🏻